### PR TITLE
cilium: add option to configure service annotation-based dispatch

### PR DESCRIPTION
**Pull request by borkmann**
(see commit desc)

### CODE REVIEW

**Review by learnitall — APPROVED** [review]
:rocket:

**Review by brb — APPROVED** [review]
Nice!

File: bpf/lib/common.h [hunk 1]
```diff
@@ -896,13 +896,14 @@ enum {
...
+	SVC_FLAG_LOCALREDIRECT  = (1 << 0),  /* Local redirect service */
 	SVC_FLAG_NAT_46X64      = (1 << 1),  /* NAT-46/64 entry */
 	SVC_FLAG_L7LOADBALANCER = (1 << 2),  /* tproxy redirect to local l7 loadbalancer */
-	SVC_FLAG_LOOPBACK       = (1 << 3),  /* hostport with a loopback hostIP */
+	SVC_FLAG_LOOPBACK       = (1 << 3),  /* HostPort with a loopback hostIP */
 	SVC_FLAG_INT_LOCAL_SCOPE = (1 << 4), /* internalTrafficPolicy=Local */
-	SVC_FLAG_TWO_SCOPES     = (1 << 5),  /* two sets of backends are used for external/internal connections */
+	SVC_FLAG_TWO_SCOPES     = (1 << 5),  /* Two sets of backends are used for external/internal connections */
 	SVC_FLAG_QUARANTINED    = (1 << 6),  /* Backend slot (key: backend_slot > 0) is quarantined */
+	SVC_FLAG_MODE_FLIP      = (1 << 7),  /* If bit is set, use DSR/SNAT if SNAT/DSR is default */
```
Comment (julianwiedmann) [inline bpf/lib/common.h:906]: This "flip" behaviour feels a bit complicated - wouldn't `SVC_FLAG_MODE_BYUSER_DSR` have been more straight-forward?

File: bpf/lib/nodeport.h [hunk 2]
```diff
@@ -3051,17 +3068,6 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
...
-		/* When BPF-Masquerading is off, we can skip the revSNAT path via
-		 * CILIUM_CALL_IPV4_NODEPORT_NAT_INGRESS if:
-		 * - the packet is ICMP, or
-		 * - the packet is DSR-eligible (and thus not reply traffic by
-		 *   a remote backend that would require revSNAT / revDNAT)
-		 */
-		if (!is_svc_proto || nodeport_uses_dsr4(&tuple))
-			return CTX_ACT_OK;
-#endif /* ENABLE_MASQUERADE_IPV4 */
-
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:3064]: For my understanding - we could have preserved the ICMP optimization here, correct?

**Review by julianwiedmann — COMMENTED** [review]
Late to the party, but trying to stay in the loop for when I need to touch it next :)

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (borkmann) [reply bpf/lib/nodeport.h:3064]: Hm, good point. I can restore that one.

File: bpf/lib/common.h (hunk 1 above)
Comment (borkmann) [reply bpf/lib/common.h:906]: Ack, I'll send a follow-up and change it into SVC_FLAG_MODE_BYUSER_DSR.

File: bpf/lib/nodeport.h (hunk 2 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:3064]: Is your long-term expectation that the annotation-mode will be enabled by default? Just thinking if we could have preserved more of the existing behaviour if `ENABLE_DSR_BYUSER` is not set.

But not sure if it's worth it, and how much compatibility we need to offer. I'd expect established DSR connections to get disrupted if a user suddenly disables the annotation-support?

**Pull request merged**